source/common/common/matchers.cc

```diff
@@ -78,6 +78,18 @@ bool MetadataMatcher::match(const envoy::api::v2::core::Metadata& metadata) cons
   case ProtobufWkt::Value::kBoolValue:
     return (bool_matcher_.has_value() && *bool_matcher_ == value.bool_value());
   case ProtobufWkt::Value::kListValue:
+    for (int i = 0; i < value.list_value().values_size(); ++i) {
+      const auto& lv = value.list_value().values(i);
+      if ((double_matcher_.has_value() && lv.kind_case() == ProtobufWkt::Value::kNumberValue &&
```

Comment:
I think you should also check `null_matcher_` unless you mention it in the doc that `null_matcher_` is not supported in list.

Another Nit: as the matcher is stable so you should be able to decide the expected value and type from the matcher outside the for-loop, and during the for-loop just check if `lv` has the value and type. This should make it faster.